internal/pkg/poster/poster.go

```diff
@@ -70,148 +81,146 @@ func (p *Poster) StopPoSting() {
-//		poStCandidates[i] = output.PoStInputs[i].Candidate
-//	}
-//
-//	err = p.sendPoSt(ctx, stateView, poStCandidates, output.Proof)
-//	if err != nil {
-//		log.Error("error sending fallback PoSt: ", err)
-//		return
-//	}
-//}
-//
-//func (p *Poster) sendPoSt(ctx context.Context, stateView *appstate.View, candidates []abi.PoStCandidate, proofs []abi.PoStProof) error {
-//
-//	windowedPost := &abi.OnChainPoStVerifyInfo{
-//		Candidates: candidates,
-//		Proofs:     proofs,
-//	}
-//
-//	_, workerAddr, err := stateView.MinerControlAddresses(ctx, p.minerAddr)
-//	if err != nil {
-//		return err
-//	}
-//
-//	mcid, _, err := p.outbox.Send(
-//		ctx,
-//		workerAddr,
-//		p.minerAddr,
-//		types.ZeroAttoFIL,
-//		types.NewGasPrice(1),
-//		gas.NewGas(5000),
-//		true,
-//		builtin.MethodsMiner.SubmitWindowedPoSt,
-//		windowedPost,
-//	)
-//	if err != nil {
-//		return err
-//	}
-//
-//	// wait until we see the post on chain at least once
-//	err = p.waiter.Wait(ctx, mcid, func(_ *block.Block, _ *types.SignedMessage, recp *vm.MessageReceipt) error {
-//		return nil
-//	})
-//	if err != nil {
-//		return err
-//	}
-//
-//	return nil
-//}
+func (p *Poster) doPoSt(ctx context.Context, stateView *appstate.View, deadlineIndex uint64, challengeAt abi.ChainEpoch, head block.TipSetKey) {
+	defer p.cancelPost()
 
-func (p *Poster) getProvingSet(ctx context.Context, stateView *appstate.View) ([]abi.SectorInfo, error) {
-	return consensus.NewPowerTableView(stateView, stateView).SortedSectorInfos(ctx, p.minerAddr)
+	minerID, err := address.IDFromAddress(p.minerAddr)
+	if err != nil {
+		log.Errorf("Error retrieving miner ID from address %s: %s", p.minerAddr, err)
+		return
+	}
+
+	partitions, err := stateView.MinerPartitionIndicesForDeadline(ctx, p.minerAddr, deadlineIndex)
+	if err != nil {
+		log.Errorf("Error retrieving partitions for address %s at index %d: %s", p.minerAddr, deadlineIndex, err)
+		return
+	}
+
+	// if no partitions, we're done
+	if len(partitions) == 0 {
+		return
+	}
+
+	// Some day we might want to choose a subset of partitions to prove at one time. Today is not that day.
+	sectors, err := stateView.MinerSectorInfoForDeadline(ctx, p.minerAddr, deadlineIndex, partitions)
+	if err != nil {
+		log.Errorf("error retrieving sector info for miner %s partitions at index %d: %s", p.minerAddr, deadlineIndex, err)
+		return
+	}
+
+	// if no sectors, we're done
+	if len(sectors) == 0 {
+		return
+	}
+
+	buf := new(bytes.Buffer)
+	err = p.minerAddr.MarshalCBOR(buf)
+	if err != nil {
+		log.Errorf("could not create entropy from miner address %s: %s", p.minerAddr, err)
+		return
+	}
+
+	randomness, err := p.chain.SampleChainRandomness(ctx, head, acrypto.DomainSeparationTag_WindowedPoStChallengeSeed, challengeAt, buf.Bytes())
+	if err != nil {
+		log.Errorf("could not sample chain randomness at %d: %s", challengeAt, err)
+		return
+	}
+
+	proofs, err := p.mgr.GenerateWindowPoSt(ctx, abi.ActorID(minerID), sectors, abi.PoStRandomness(randomness))
```

Comment:
You'll need to pass the on-chain miner faults and recoveries bitfields in here (or maybe a prior call when `sectors` is calculated) in order to calculate the right proof. I think that's worth doing now even though this impl doesn't yet report faults. The sector info for any faulty sector must be replaced by the info for the first non-faulty sector (see the actor code which does the same).

Reply:
We now do the set logic in the state view. This is just copied from the actor. It's a lot of logic and probably should be a method on miner state.